packages/dev/core/src/Collisions/gpuPicker.ts

```diff
@@ -0,0 +1,314 @@
+
+                    colorData[(i + 1) * 4] = r / 255.0;
+                    colorData[(i + 1) * 4 + 1] = g / 255.0;
+                    colorData[(i + 1) * 4 + 2] = b / 255.0;
+                    colorData[(i + 1) * 4 + 3] = 1.0;
+                    id++;
+                }
+
+                const buffer = new VertexBuffer(engine, colorData, VertexBuffer.ColorKind, false, false, 4, true);
+                (mesh as Mesh).setVerticesBuffer(buffer, true);
+            } else {
+                this._idColors[mesh.uniqueId] = Color3.FromInts(r, g, b);
+            }
+        }
+    }
+
+    /**
+     * Execute a picking operation
+     * @param x defines the X coordinates where to run the pick
+     * @param y defines the Y coordinates where to run the pick
+     * @param scene defines the scene to pick from
+     * @param disposeWhenDone defines a boolean indicating we do not want to keep resources alive
+     * @returns A promise with the picking results
+     */
+    public pickAsync(x: number, y: number, scene: Scene, disposeWhenDone = true): Promise<Nullable<AbstractMesh>> {
+        const engine = scene.getEngine();
+        const rttSizeW = engine.getRenderWidth();
+        const rttSizeH = engine.getRenderHeight();
+
+        if (!this._readbuffer) {
+            this._readbuffer = new Uint8Array(engine.isWebGPU ? 256 : 4); // Because of block alignment in WebGPU
+        }
+
+        this._meshRenderingCount = 0;
+        // Ensure ints
+        x = x >> 0;
+        y = y >> 0;
+
+        if (x < 0 || y < 0 || x >= rttSizeW || y >= rttSizeH) {
+            return Promise.resolve(null);
+        }
+
+        // Invert Y
+        y = rttSizeH - y;
+
+        if (!this._pickingTexure) {
+            this._createRenderTarget(scene, rttSizeW, rttSizeH);
+        } else {
+            const size = this._pickingTexure.getSize();
+
+            if (size.width !== rttSizeW || size.height !== rttSizeH || this._cachedScene !== scene) {
+                this._pickingTexure.dispose();
+                this._createRenderTarget(scene, rttSizeW, rttSizeH);
+            }
+        }
+
+        if (!this._cachedScene || this._cachedScene !== scene) {
+            this._createColorMaterial(scene);
+        }
+
+        this._cachedScene = scene;
+        scene.customRenderTargets.push(this._pickingTexure!);
+        this._pickingTexure!.renderList = [];
+        this._pickingTexure!.clearColor = new Color4(0, 0, 0, 0);
+
+        // We need to give every mesh an unique color (when there is no picking list)
+        this._pickingTexure!.onBeforeRender = () => {
+            if (!this._userDefinedList) {
+                this._pickableMeshes = scene.meshes.filter((m) => (m as Mesh).geometry && m.isPickable && (m as Mesh).onBeforeRenderObservable) as Mesh[];
+            }
+
+            for (let index = 0; index < this._pickableMeshes.length; index++) {
+                const mesh = this._pickableMeshes[index];
+                if (!mesh.isAnInstance) {

```

Comment:
If mesh is an instance, instances won't correctly be picked when `!this._userDefinedList`.

You should probably factor out the core code of `setPickingList` so that it can be reused here when `!this._userDefinedList`. However, we should try not to recreate a vertex buffer each time `pickAsync` is called...

Maybe we should simplify everything by not having a `onBeforeRender` but requires the user to call `setPickingList` before doing any picking:
* if `null` is passed, we take all meshes from the scene as the pickable mesh list
* else we use the passed list

After all, most of the time, the list of pickable meshes won't change between consecutive picking (and if it changes, it would be the user responsibility to call `setPickingList` again).